pkg/controller/ippool/controller.go

```diff
@@ -375,6 +377,10 @@ func (h *Handler) MonitorAgent(ipPool *networkv1.IPPool, status networkv1.IPPool
 		return status, err
 	}
 
+	if agentPod.GetUID() != ipPool.Status.AgentPodRef.UID || agentPod.Spec.Containers[0].Image != ipPool.Status.AgentPodRef.Image {
```

Comment:
Maybe compare the container image with the agent image name stored in the controller?

```suggestion
	if agentPod.GetUID() != ipPool.Status.AgentPodRef.UID || agentPod.Spec.Containers[0].Image !=  h.agentImage.String() {
```


<details>
<summary>How I tested</summary>

I check out the PR with the `test1` branch, build and publish to my repo, and edit the values.file:
```
image:
  repository: bk201z/harvester-vm-dhcp-controller
  pullPolicy: Always
  # Overrides the image tag whose default is the chart appVersion.
  tag: "test1-head"

agent:
  image:
    repository: bk201z/harvester-vm-dhcp-agent
    pullPolicy: Always
    tag: "test1-head"

webhook:
  replicaCount: 1
  image:
    repository: bk201z/harvester-vm-dhcp-webhook
    pullPolicy: Always
    tag: "test1-head"
```

Install chart and create an IPPool
```
helm upgrade --install harvester-vm-dhcp-controller ./chart --namespace=harvester-system --create-namespace
```

Check out branch `test2` and re-publish, and edit the values file again:
```
image:
  repository: bk201z/harvester-vm-dhcp-controller
  pullPolicy: Always
  # Overrides the image tag whose default is the chart appVersion.
  tag: "test2-head"

agent:
  image:
    repository: bk201z/harvester-vm-dhcp-agent
    pullPolicy: Always
    tag: "test2-head"

webhook:
  replicaCount: 1
  image:
    repository: bk201z/harvester-vm-dhcp-webhook
    pullPolicy: Always
    tag: "test2-head"
```

And then upgrade the chart:
```
helm upgrade --install harvester-vm-dhcp-controller ./chart --namespace=harvester-system --create-namespace
```

The agent pod seems not to be restarted.

</details>

Reply:
Since the agent image information is an internal controller variable (not in an object's spec), we need to reflect it on the object status first (done by the `ippool-register` reconcile loop). Then, rely on the other reconciliation loop to sync the status with the real pod.

It's the `ippool-agent-monitor` reconcile loop's mission to ensure that the running agent pod does not drift away. If it finds the agent pod it tracks does not match what's recorded in the AgentPodRef, it deletes the pod.

Thank you for pointing out there's a bug that the AgentPodRef is not updated if the agent pod already exists. I've updated the relevant part. Please take a look :)